### PR TITLE
HDDS-4553. ChunkInputStream should release buffer as soon as last byte in the buffer is read

### PR DESCRIPTION
**Pull request by hanishakoneru**
## What changes were proposed in this pull request?

We currently wait for reading up till the Chunk EOF before releasing the buffers in ChunkInputStream (or when the stream is closed). Let's say a client reads first 3 MB of a chunk  of size 4MB and does not close the stream immediately. This would lead to the 3MB of data being cached in the ChunkInputStream buffers till the stream is closed.

After HDDS-4552, a chunk read from DN would return the chunk data as an array of ByteBuffers. After each ByteBuffer is read, it can be released. This would greatly help with optimizing memory usage of ChunkInputStream.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4553

## How was this patch tested?

Added unit tests

### CODE REVIEW

**Comment by hanishakoneru** [comment]
@adoroszlai, I fixed the review comments you posted in HDDS-4552 here. Please take a look. Thanks.

**Comment by adoroszlai** [comment]
> I fixed the review comments you posted in [HDDS-4552](https://issues.apache.org/jira/browse/HDDS-4552) [here](https://github.com/apache/ozone/pull/2062/commits/42f29eba3955a5aef36c26d1b17473d78abc402b)

Thanks.  That part looks good.  I'll check the rest later.

**Comment by umamaheswararao** [comment]
@hanishakoneru Thank you for the patch. I have posted few minor comments. Please check when you get time. Thanks

**Comment by hanishakoneru** [comment]
Thank you @umamaheswararao for the review. I have addressed your comments in the latest commit.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ChunkInputStream.java [hunk 1]
```diff
@@ -628,12 +645,38 @@ private boolean chunkStreamEOF() {
     }
   }
 
+
+  /**
+   * Release a buffers upto the given index.
```
Comment (umamaheswararao) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ChunkInputStream.java:650]: Tiny nit: make sure to change it before commit
Release a buffers --> Release the buffers

**Review by umamaheswararao — APPROVED** [review]
Thank latest changes looks good to me. 
Before please make sure to update tiny nit comment.

**Review by bshashikant — APPROVED** [review]
Looks good. Thanks @hanishakoneru for working on this.

**Comment by hanishakoneru** [comment]
Thank you @umamaheswararao and @bshashikant for the reviews.

**Pull request merged**